src/sat/sat_clause.cpp

```diff
@@ -123,48 +123,80 @@ namespace sat {
 
     clause_allocator::clause_allocator():
         m_allocator("clause-allocator") {
-#ifdef _AMD64_
+#if defined(_AMD64_) 
         m_num_segments = 0;
+#if defined(Z3DEBUG)
+        m_overflow_valid = false;
+#endif
 #endif
     }
 
     clause * clause_allocator::get_clause(clause_offset cls_off) const {
-#ifdef _AMD64_
+#if defined(_AMD64_) 
+#if defined (Z3DEBUG)
+        clause const* result;
+        if (m_overflow_valid && m_cls_offset2ptr.find(cls_off, result)) {
+            return const_cast<clause*>(result);
+        }
+#endif
         return reinterpret_cast<clause *>(m_segments[cls_off & c_aligment_mask] + (static_cast<size_t>(cls_off) & ~c_aligment_mask));
 #else
         return reinterpret_cast<clause *>(cls_off);
 #endif
     }
 
-#ifdef _AMD64_
-    unsigned clause_allocator::get_segment(size_t ptr) {
+#if defined(_AMD64_) 
+    unsigned clause_allocator::get_segment(clause const* cls) {
+        size_t ptr = reinterpret_cast<size_t>(cls);
```

Comment:
This looks like taking the size of the pointer. If so, it would feel more correct to have `auto ptr = reinterpret_cast<std::uintptr_t>(cls);`, I believe. Some more information at http://stackoverflow.com/questions/153065/converting-a-pointer-into-an-integer.